### PR TITLE
add skip navigation link

### PR DESCRIPTION
**Pull request by annieyw**
A soft nav should auto focus on the content and a hard nav should have the next focus element as skip link.
A hard nav is a full page load (e.g. when coming to the app from another site), a soft nav is a route change within the app. These terms are further explained in https://dzone.com/articles/quick-primer-to-single-page-application-spa-and-tr

### CODE REVIEW

**Comment by atscott** [comment]
I think the headbutting headers was merged yesterday, right? Can you drop that commit from this PR and update the commit message for the second one to be more descriptive?

File: src/app/pages/component-sidenav/component-sidenav.html [hunk 1]
```diff
@@ -11,7 +11,7 @@
   <div class="docs-component-sidenav-content">
     <component-page-header (toggleSidenav)="toggleSidenav(sidenav)"></component-page-header>
     <div class="docs-component-sidenav-inner-content">
-      <main class="docs-component-sidenav-body-content">
+      <main class="docs-component-sidenav-body-content" id="main-content" role="main" aria-label="Side navigation bar">
```
Comment (atscott) [inline src/app/pages/component-sidenav/component-sidenav.html:14]: Is this the correct label? It looks like this is the main body contents, not the navigation bar.

File: src/app/material-docs-app.html [hunk 2]
```diff
@@ -1,2 +1,5 @@
+<div id="skip-link-wrapper">
```
Comment (mmalerba) [inline src/app/material-docs-app.html:1]: We like to uses classes over ids when possible because the same class can appear on more than 1 element and CSS rules targeted at classes are less specific than rules targeted at ids.

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 3]
```diff
@@ -9,16 +11,32 @@ let lastTimeoutId = -1;
...
-    lastTimeoutId = window.setTimeout(() => this.el.nativeElement.focus({preventScroll: true}),
-      100);
+    let previousUrl = window.location.hostname;
+    this.router.events
+      .pipe(filter((event: Event) => event instanceof NavigationEnd))
+      .subscribe(() => {
+        const afterUrl = window.location.hostname;
+        // only want to focus on the element if it the navigation was a soft nav,
+        // otherwise first focusable element should be the skip link
+        if (isSoftNav(previousUrl)) {
```
Comment (atscott) [inline src/app/shared/navigation-focus/navigation-focus.ts:24]: Can you determine this from the `id` on the `RouterEvent` https://angular.io/api/router/RouterEvent? A hard nav would have an id of 0, right?

File: src/app/material-docs-app.scss [hunk 4]
```diff
@@ -26,6 +26,28 @@ material-docs-app > guide-viewer {
   overflow-y: auto;
 }
 
+#skip-link-wrapper {
+  z-index: 10;
+  position: fixed;
```
Comment (mmalerba) [inline src/app/material-docs-app.scss:31]: Should this be moved inside the `app-navbar`? That would probably make it so we don't need the `position: fixed` and `z-index`, since the app bar already has fixed positioning

File: src/app/material-docs-app.scss [hunk 5]
```diff
@@ -26,6 +26,28 @@ material-docs-app > guide-viewer {
...
+  top: 15px;
+  width: 16%;
+  margin: 0 42% 0 42%;
+  text-align: center;
+  padding: 0.2em 0 0.2em 0;
+  background: white;
+  border-radius: 20px;
+  opacity: 0;
```
Comment (mmalerba) [inline src/app/material-docs-app.scss:39]: Is `opacity: 0` sufficient to prevent the element from receiving click events?

File: src/app/material-docs-app.ts [hunk 6]
```diff
@@ -24,9 +24,13 @@ export class MaterialDocsApp {
         if (!isNavigationWithinComponentView(previousRoute, urlAfterRedirects)) {
           resetScrollPosition();
         }
-
         previousRoute = urlAfterRedirects;
         ga.locationChanged(urlAfterRedirects);
+
+        // set skip link
+        if(!router.url.endsWith('#main-content')) {
+          this.skipLinkTarget = router.url + "#main-content";
```
Comment (mmalerba) [inline src/app/material-docs-app.ts:32]: I could see this breaking for a URL like `.../stuff#not-the-main-content`. I think you should just be able to set the `href` to `#main-content` and the rest of the URL is assumed

File: src/app/pages/guide-list/guide-list.html [hunk 7]
```diff
@@ -1,8 +1,8 @@
 <header class="docs-primary-header">
-  <h1>Guides</h1>
+  <h1 focusOnNavigation>Guides</h1>
```
Comment (mmalerba) [inline src/app/pages/guide-list/guide-list.html:2]: It seems weird to me that the auto-focused element and the element the skip link takes you to are not the same

File: src/app/pages/homepage/homepage.html [hunk 8]
```diff
@@ -1,5 +1,5 @@
 <header class="docs-header-background"
-        [class.is-next-version]="isNextVersion">
+        [class.is-next-version]="isNextVersion" id="main-content" role="main" aria-label="Get started">
```
Comment (mmalerba) [inline src/app/pages/homepage/homepage.html:2]: On other pages the `<main>` element is the `#main-content`, is this a mistake?

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 9]
```diff
@@ -9,16 +11,32 @@ let lastTimeoutId = -1;
...
+          lastTimeoutId = window.setTimeout(() => this.el.nativeElement.focus({preventScroll: true}),
+            100);
+        }
+        previousUrl = afterUrl;
+      });
   }
 }
 
+function isSoftNav(url: string) {
+  const withinApp = /(localhost|material.angular.io)/;
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:36]: This is brittle, I would just compare the old and new hostname with `===`.

@atscott do you know if there's a more "Angular" way of distinguishing soft nav from hard nav?

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 10]
```diff
@@ -9,16 +11,32 @@ let lastTimeoutId = -1;
...
 
-  constructor(private el: ElementRef) {}
+  constructor(private el: ElementRef, private router: Router) {}
 
   ngOnInit() {
-    clearTimeout(lastTimeoutId);
-    // 100ms timeout is used to allow the page to settle before moving focus for screen readers.
-    lastTimeoutId = window.setTimeout(() => this.el.nativeElement.focus({preventScroll: true}),
-      100);
+    let previousUrl = window.location.hostname;
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:17]: Technically `previousHostName` which is not the same as a URL

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 11]
```diff
@@ -9,16 +11,32 @@ let lastTimeoutId = -1;
...
+    this.router.events
+      .pipe(filter((event: Event) => event instanceof NavigationEnd))
+      .subscribe(() => {
+        const afterUrl = window.location.hostname;
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:21]: It looks like this variable is only used once, might as well just inline it

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 3 above)
Comment (mmalerba) [reply src/app/shared/navigation-focus/navigation-focus.ts:24]: I figured there had to be some way, but I had no idea what it was. Also, I would never guess that that's what 0 means from the docs

Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:24]: Yea, it's not immediately apparent. Using the `id` won't work if the initial navigation results in a redirect from a guard calling `router.navigate` or returning `UrlTree`. The initial navigation would then be cancelled and the following would have `id` 1... Still trying to think of a more "Angular" way to do this that works in all cases. 
I'm not 100% confident that the current way would work in the redirect case either. @vxlydia - do you think you can you test this out?

File: src/app/pages/homepage/homepage.html (hunk 8 above)
Comment (vxlydia) [reply src/app/pages/homepage/homepage.html:2]: I wasn't sure whether to put it here or in main because there are 2 "Get Started" buttons and this one was the first one

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 11 above)
Comment (vxlydia) [reply src/app/shared/navigation-focus/navigation-focus.ts:21]: Changed to using `id` from `RouterEvent`

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 3 above)
Comment (vxlydia) [reply src/app/shared/navigation-focus/navigation-focus.ts:24]: I tried checking if `id` is 0 and it seems so work. I'm not sure how to test for the redirect case, could you point me to an example?

Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:24]: The only guard I could find was here: https://github.com/angular/material.angular.io/blob/master/src/app/pages/component-sidenav/component-sidenav-can-load-guard.ts and it does call `router.navigate`. So to test out this cancellation/redirect behavior, navigate to `localhost:4200/nonsenseroute`.

File: src/app/pages/component-sidenav/component-sidenav.html (hunk 1 above)
Comment (Splaktar) [reply src/app/pages/component-sidenav/component-sidenav.html:14]: A `main` element also doesn't need to have an additional `role="main"` defined.

**Comment by Splaktar** [comment]
Can you please add a link to the PR description to a site/blog/documentation that defines and describes this soft-nav and hard-nav pattern and the vocabulary involved? Maybe there's a WAI-ARIA practice that covers this?

File: src/app/pages/homepage/homepage.html (hunk 8 above)
Comment (Splaktar) [reply src/app/pages/homepage/homepage.html:2]: It doesn't seem expected or ideal for a `header` element to have `role="main"`.

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 9 above)
Comment (Splaktar) [reply src/app/shared/navigation-focus/navigation-focus.ts:36]: We'll want this to work with the following as well
- https://next.material.angular.io/
- https://v6.material.angular.io/ (also v5, v7, v8, etc)

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 12]
```diff
@@ -9,16 +11,32 @@ let lastTimeoutId = -1;
...
+        // only want to focus on the element if it the navigation was a soft nav,
```
Comment (Splaktar) [inline src/app/shared/navigation-focus/navigation-focus.ts:22]:
```suggestion
        // only want to focus on the element if the navigation was a soft nav,
```

**Comment by mmalerba** [comment]
@Splaktar This post talks about it a bit: https://dzone.com/articles/quick-primer-to-single-page-application-spa-and-tr

TL;DR a hard nav is a full page load (e.g. when coming to the app from another site), a soft nav is a route change within the app. 

I couldn't really find anything talking about how a skip link should work with soft nav, but we thought it makes sense to put the focus at the top on the first page load so people can explore the nav links, and then put the focus at the content the user is trying to navigate to on subsequent route changes

File: src/app/material-docs-app.scss (hunk 5 above)
Comment (mmalerba) [reply src/app/material-docs-app.scss:39]: Yeah just checked this, `opacity:0` is not sufficient. We have a class in the CDK called `cdk-visually-hidden` that should handle this, I would just apply that class to the element

File: src/app/pages/homepage/homepage.html (hunk 8 above)
Comment (mmalerba) [reply src/app/pages/homepage/homepage.html:2]: Ok this makes sense after I looked at the page. Like Michael said, I would just remove the `role="main"`

**Comment by Splaktar** [comment]
OK, that makes sense. 

On https://material.angularjs.org (AngularJS Material), on a hard nav we don't have any focus, in this case the 2nd tabbable element is the skip link and the third is the nav.

For soft nav, we focus the header text in the toolbar (i.e. "Demos > Checkbox", [Checkbox Demo link](https://material.angularjs.org/latest/demo/checkbox)) to allow the user to quickly get into the content.

File: src/app/shared/navbar/navbar.html [hunk 13]
```diff
@@ -1,6 +1,11 @@
 <!-- TODO: figure out if the <nav> should go inside of a <header> element. -->
 <nav class="docs-navbar-header" aria-label="Top Toolbar"
      [class.is-next-version]="isNextVersion">
+  <div class="skip-link-wrapper" [class.cdk-visually-hidden]=skipLinkHidden>
+    <a [href]=skipLinkTarget id="skip-link" (focus)="setSkipLinkVisible()" (blur)="setSkipLinkInvisible()">
```
Comment (mmalerba) [inline src/app/shared/navbar/navbar.html:5]: is the ID needed for some reason? I don't see any code that references it

Comment (mmalerba) [inline src/app/shared/navbar/navbar.html:5]: nit: rather than having 2 method you could just have one that takes an argument, or 0 methods: `(focus)="skipLinkHidden = false"`

File: src/app/shared/navbar/navbar.ts [hunk 14]
```diff
@@ -19,6 +20,17 @@ const SECTIONS_KEYS = Object.keys(SECTIONS);
...
   isNextVersion = location.hostname.startsWith('next.material.angular.io');
+  skipLinkTarget: string;
+  skipLinkHidden = true;
+  constructor(router: Router) {
+    router.events
+      .pipe(filter((event: Event) => event instanceof NavigationEnd))
+      .subscribe(() => {
+        // set skip link
+        const baseUrl = router.url.split('#')[0];
+        this.skipLinkTarget = `${baseUrl}#main-content`;
```
Comment (mmalerba) [inline src/app/shared/navbar/navbar.ts:31]: can we call this `skipLinkHref`, `target` is actually a different attribute on `<a>` tags that tells it whether to open in the same window or a new one

File: src/app/material-docs-app.ts (hunk 6 above)
Comment (vxlydia) [reply src/app/material-docs-app.ts]: setting `href` directly to `#main-content` will cause the page to redirect to the homepage. 
a similar issue is addressed in [doc-viewer.ts](https://github.com/angular/material.angular.io/blob/master/src/app/shared/doc-viewer/header-link.ts)

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 15]
```diff
@@ -9,16 +11,29 @@ let lastTimeoutId = -1;
...
-    clearTimeout(lastTimeoutId);
-    // 100ms timeout is used to allow the page to settle before moving focus for screen readers.
-    lastTimeoutId = window.setTimeout(() => this.el.nativeElement.focus({preventScroll: true}),
-      100);
+    this.router.events
+      .pipe(filter((event: Event): event is RouterEvent => event instanceof RouterEvent))
+      .subscribe((routerEvent: RouterEvent) => {
+        // only want to focus on the element if the navigation was a soft nav,
+        // otherwise first focusable element should be the skip link
+        if (!(routerEvent.id === 0)) {
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:22]: can we make this `if(isSoftNav(routerEvent))`, easier to follow the logic that way

File: src/app/shared/navbar/navbar.html [hunk 16]
```diff
@@ -1,6 +1,11 @@
 <!-- TODO: figure out if the <nav> should go inside of a <header> element. -->
 <nav class="docs-navbar-header" aria-label="Top Toolbar"
      [class.is-next-version]="isNextVersion">
+  <div class="skip-link-wrapper" [class.cdk-visually-hidden]=skipLinkHidden>
```
Comment (atscott) [inline src/app/shared/navbar/navbar.html:4]: Style nit: always use quotes for the attribute values, i.e. `[class.cdk-visually-hidden]="skipLinkHidden"` and `[href]="skipLinkTarget"`

File: src/app/shared/navbar/navbar.html (hunk 13 above)
Comment (vxlydia) [reply src/app/shared/navbar/navbar.html:5]: added that plus `(blur)="skipLinkHidden = true"` otherwise the skip link would stay there

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 17]
```diff
@@ -9,16 +11,34 @@ let lastTimeoutId = -1;
...
+  private currentLocation: URL;
+
+  constructor(private el: ElementRef, private router: Router) {}
 
   ngOnInit() {
-    clearTimeout(lastTimeoutId);
-    // 100ms timeout is used to allow the page to settle before moving focus for screen readers.
-    lastTimeoutId = window.setTimeout(() => this.el.nativeElement.focus({preventScroll: true}),
-      100);
+    this.currentLocation = new URL(window.location.href);
```
Comment (atscott) [inline src/app/shared/navigation-focus/navigation-focus.ts:20]: I'm confused about how this works. Wouldn't the location *have* to be `localhost`/`material.io` at this point? I don't know if you're able to detect "hard" navs with this method. Maybe I'm missing something though.

Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:20]: There's already code in the `MaterialDocsApp` that tracks `NavigationEnd` events. Can you maybe add another that's something like:
```
router.events
      .pipe(
                filter((event: Event) => event instanceof NavigationEnd)),
                skip(1),
                take(1),
      ).subscribe(() => globalState.internallyNavigated = true);
```

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 18]
```diff
@@ -9,16 +11,34 @@ let lastTimeoutId = -1;
...
+    this.router.events.pipe(filter((event: Event) => event instanceof NavigationEnd))
+      .subscribe(() => {
+        this.previousLocation = this.currentLocation;
+        this.currentLocation = new URL(window.location.href);
+        if (this.currentLocation.hash) {
+          clearTimeout(lastTimeoutId);
+          // 100ms timeout is used to allow the page to settle before moving focus for screen readers.
+          lastTimeoutId = window.setTimeout(() => document.getElementById(this.currentLocation.hash)!.focus(), 100);
```
[the remark this answers is not in the document]
Comment (mmalerba) [reply src/app/shared/navigation-focus/navigation-focus.ts:28]: I think we had some problems at some point in the past where the element with the ID in question was late loaded and so it just linked to the top. We can test commenting this out and see if links within the component overview still work

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 17 above)
Comment (mmalerba) [reply src/app/shared/navigation-focus/navigation-focus.ts:20]: Does the router not fire events when I navigate from e.g. about:blank --> m.a.io/components/autocomplete/overview? I was under the impression that it did, but I admittedly don't know much about the router

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:28]: Ahhh, I see. That could be the case.

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 17 above)
Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:20]: It does fire events for the initial navigation. about:blank --> m.a.io/components/autocomplete/overview would fire all the navigation events. I don't think I'm answering the question you're asking though... By the time any JS runs on the page, the URL is already m.a.io/components/autocomplete/overview

Comment (mmalerba) [reply src/app/shared/navigation-focus/navigation-focus.ts:20]: Oh, yeah I see what you're saying. I think for this method to work properly, the previousUrl would have to be globally saved and initially null, then we update it after each navigation. I think Annie is planning to switch this back to using the id property to detect soft navs like you originally suggested anyways

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 19]
```diff
@@ -1,24 +1,57 @@
...
+  constructor(private el: ElementRef, private router: Router) {
+    // subscribing to events in constructor to catch the correct NavigationEnd/ResolveEnd from previous page
+    // and determine whether the element in question needs to be focused
+    // switching between different components/cdk produces a ResolveEnd instead of NavigationEnd event
+    // the focus is set in ngOnInit to account for late loaded content
+    this.subscriptions.add(
+      this.router.events
+        .pipe(
+          filter((event: Event) => event instanceof NavigationEnd || event instanceof ResolveEnd),
+          map(navEnd => navEnd as NavigationEnd))
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:27]: This map is unnecessary, it's just serving as an expensive type cast. Also it seems like maybe a lie since it could be a `ResolveEnd` - why do we want `ResolveEnd` in there? That seems not right

You could maybe try using this alternate signature of filter (notice 2 type parameters):
```ts
function filter<T, S extends T>(predicate: (value: T, index: number) => value is S,
                                       thisArg?: any): OperatorFunction<T, S>;
```

In the context of your code it would look something like:
```ts
filter<RouterEvent, NavigationEnd>(...)
```

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 20]
```diff
@@ -1,24 +1,57 @@
...
-export class NavigationFocus implements OnInit {
+export class NavigationFocus implements OnInit, OnDestroy {
   @HostBinding('tabindex') role = '-1';
 
-  constructor(private el: ElementRef) {}
+  private softNavFocus = false;
+  private subscriptions = new Subscription();
+
+  constructor(private el: ElementRef, private router: Router) {
+    // subscribing to events in constructor to catch the correct NavigationEnd/ResolveEnd from previous page
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:19]: I would change this comment to:
```
// We need to subscribe in the constructor in order to catch the `NavigationEnd` event 
// from navigating from the previous page to this page. This event fires before 
// `ngOnInit` is called. However we must wait until `ngOnInit` for the element to be 
// rendered before we can focus it
```

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 21]
```diff
@@ -1,24 +1,57 @@
...
+      this.router.events
+        .pipe(
+          filter((event: Event) => event instanceof NavigationEnd || event instanceof ResolveEnd),
+          map(navEnd => navEnd as NavigationEnd))
+      .subscribe((navEnd: NavigationEnd) => {
+        const currentLocation = new URL(window.location.href);
+        if (!currentLocation.hash && isSoftNav(navEnd)) {
+          this.softNavFocus = true;
+          // navigating between CDK/Components pages only runs ngOnInit once so the focus must also be set here
+          this.el.nativeElement.focus({preventScroll: true});
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:33]: I thought we were waiting until `ngOnInit` to focus?

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 22]
```diff
@@ -1,24 +1,57 @@
...
 
   ngOnInit() {
-    clearTimeout(lastTimeoutId);
-    // 100ms timeout is used to allow the page to settle before moving focus for screen readers.
-    lastTimeoutId = window.setTimeout(() => this.el.nativeElement.focus({preventScroll: true}),
-      100);
+    if (this.softNavFocus) {
+      // 100ms timeout is used to allow the page to settle before moving focus for screen readers.
+      clearTimeout(lastTimeoutId);
+      lastTimeoutId = window.setTimeout(()=>this.el.nativeElement.focus({preventScroll: true}), 100);
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:42]: I wonder if we could eliminate the `setTimeout` by doing the focus in `ngAfterViewInit` instead. Might be worth testing

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 23]
```diff
@@ -1,24 +1,57 @@
...
+    }
+  }
+
+  ngOnDestroy(): void {
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:46]: nit: don't need to explicitly put `: void`

File: src/app/shared/navigation-focus/navigation-focus.ts [hunk 24]
```diff
@@ -1,24 +1,57 @@
...
+    this.subscriptions.unsubscribe();
   }
 }
 
+function isSoftNav(navigationEnd: NavigationEnd) {
+  return navigationEnd.id !== 1;
```
Comment (mmalerba) [inline src/app/shared/navigation-focus/navigation-focus.ts:52]: I thought @atscott said `0` was a hardnav? How did you get `1`, is that just from experimenting?

File: src/app/shared/navbar/navbar.html [hunk 25]
```diff
@@ -1,6 +1,11 @@
 <!-- TODO: figure out if the <nav> should go inside of a <header> element. -->
 <nav class="docs-navbar-header" aria-label="Top Toolbar"
      [class.is-next-version]="isNextVersion">
+  <div class="skip-link-wrapper" [class.cdk-visually-hidden]="skipLinkHidden" >
```
[the remark this answers is not in the document]
Comment (atscott) [reply src/app/shared/navbar/navbar.html:4]: I tried investigating this as well and couldn't find a way to make `routerLink` work. In general, that would be more desirable, but because the `NavBar` exists at the root, I couldn't figure out how to get the navigation to be relative to the final activated component rather than the root. Otherwise, you could use `<a [routerLink]="[]" fragment="main-content"></a>` or something similar to update the fragment.

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 24 above)
Comment (vxlydia) [reply src/app/shared/navigation-focus/navigation-focus.ts:52]: I confirmed with him and he said 0 is actually filtered out so it should be 1
https://github.com/angular/angular/blob/master/packages/router/src/router.ts#L448

Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:52]: Yea, sorry about that. The `initialNavigation` would be `1`, as the `id` is incremented from `0` to `1` before calling `transitions.next`: https://github.com/angular/angular/blob/master/packages/router/src/router.ts#L1151. To be honest, I don't even know if that filter is necessary in the router pipe (it may have been at some point in the past).

Comment (mmalerba) [reply src/app/shared/navigation-focus/navigation-focus.ts:52]: Ah ok, do you think this is something worth having a more official API for? If so maybe @vxlydia can send a PR for it

File: src/app/pages/component-viewer/component-api.html [hunk 26]
```diff
@@ -1,5 +1,5 @@
 <ng-container *ngIf="componentViewer.componentDocItem | async; let docItem">
-  <span class="cdk-visually-hidden" tabindex="-1" #initialFocusTarget>
+  <span class="cdk-visually-hidden" tabindex="-1" #initialFocusTarget focusOnNavigation>
```
Comment (atscott) [inline src/app/pages/component-viewer/component-api.html:2]: Can we remove this and use `@ViewChild(FocusOnNavigation)` instead in `ComponentBaseView`?

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 24 above)
Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:52]: That could be useful. Maybe something simple like: "each navigation has a unique id and is available on the RouterEvents" and potentially something about how the `id` is a number that's incremented. I'd be wary of making any promises about the meaning of the id beyond that.

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 21 above)
Comment (vxlydia) [reply src/app/shared/navigation-focus/navigation-focus.ts:33]: navigating between the CDK/Components page only runs ngAfterViewInit once because it's the same component (I believe) so we need to track navigating between them through the router events

File: src/app/pages/component-sidenav/component-sidenav.ts [hunk 27]
```diff
@@ -39,6 +39,7 @@ import {StackBlitzButtonModule} from '../../shared/stack-blitz';
 import {SvgViewerModule} from '../../shared/svg-viewer/svg-viewer';
 import {MatDrawerToggleResult} from '@angular/material/sidenav/drawer';
 import {MatListModule} from '@angular/material/list';
+import {NavigationFocusModule} from "../../shared/navigation-focus/navigation-focus";
```
[the remark this answers is not in the document]
Comment (vxlydia) [reply src/app/pages/component-sidenav/component-sidenav.ts]: I figured it out :)

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 22 above)
Comment (vxlydia) [reply src/app/shared/navigation-focus/navigation-focus.ts]: Seemed to work. Removed the relevant vars for timeout

File: src/app/pages/component-viewer/component-api.html (hunk 26 above)
Comment (vxlydia) [reply src/app/pages/component-viewer/component-api.html]: I'm not sure if this is what you meant but  I removed the focusOnNavigation directives on the api/over/example pages and now just have 1 on the base component viewer

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 19 above)
Comment (vxlydia) [reply src/app/shared/navigation-focus/navigation-focus.ts]: Removed map and changed to just using `ResolveEnd`

File: src/app/pages/component-viewer/component-api.html (hunk 26 above)
Comment (atscott) [reply src/app/pages/component-viewer/component-api.html]: Oh, that's a good change as well! What I meant was to remove the `#initialFocusTarget` and change `@ViewChild('initialFocusTarget') focusTarget: ElementRef;` in `ComponentBaseView` to `@ViewChild(FocusOnNavigation, {read: ElementRef}) focusTarget: ElementRef;` since we have a directive now that we can use as the selector and don't really need the 'initialFocusTarget' ref. It's not really necessary though, just a nit.

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 24 above)
Comment (mmalerba) [reply src/app/shared/navigation-focus/navigation-focus.ts:53]: Yeah that was kind of my point, making an assumption about the id _is_ hacky, you already pointed out that the filter might be unnecessary and if we remove it we'd need to change this check to compare to `0` instead of `1`. I was thinking more like adding a property to the event that can be more reliably checked against in this type of situation (I'm sure there's many ways we could accomplish this, but for example we could add `{from: Route|null, to: Router|null}` - that way a simple `if(event.from)` would tell us if its a softnav)

Comment (atscott) [reply src/app/shared/navigation-focus/navigation-focus.ts:49]: Oh, sorry - I read that as "official docs" rather than "official API". Yes, there are a lot of holes in the router APIs but I can't necessarily recommend spending too much time on new features since the plan for the future of the router is unclear. Adding more API surface that we have to maintain through a potential refactor may be undesirable.

That said, I would absolutely file an issue for this and I can triage it as a feature request. Then we can also see if it gains community support.

File: src/app/shared/navigation-focus/navigation-focus.ts (hunk 22 above)
Comment (vxlydia) [reply src/app/shared/navigation-focus/navigation-focus.ts]: Actually it didn't work for this specific case (nav from non cdk/component page to cdk/component page). Kept setting the focus in ngOnInit with timeout

**Comment by vxlydia** [comment]
navigation between pages should be functional now; PTAL

**Comment by mmalerba** [comment]
Looking pretty good, couple things I noticed on the demo:
- There's an ugly outline on the component that's focused after following the skip-link. I think you can remove it with this CSS:
`#main-content { outline: none; }`
- When you navigate to any of the guides (e.g. http://localhost:4200/guide/getting-started) the skiplink doesn't work because there is nothing with the id `main-content`. This is the fragility I mentioned in one of the comments above. We can come up with a more robust solution in a future PR, for now lets just make sure these pages have an element with the right id.
- When I follow the skiplink on http://localhost:4200/components/autocomplete/overview it takes me past the tab headers, it should probalby land just before the tab headers so I can go to the overview, api, or examples easily

File: src/app/pages/component-viewer/component-viewer.ts [hunk 28]
```diff
@@ -88,7 +89,7 @@ export class ComponentBaseView implements OnInit, OnDestroy {
   ngOnInit() {
     this.componentViewer.componentDocItem.pipe(takeUntil(this.destroyed)).subscribe(() => {
       // 100ms timeout is used to allow the page to settle before moving focus for screen readers.
-      setTimeout(() => this.focusTarget.nativeElement.focus({preventScroll: true}), 100);
+      setTimeout(() => this.focusTarget.nativeElement.id = 'main-content', 100);
```
Comment (mmalerba) [inline src/app/pages/component-viewer/component-viewer.ts:92]: We probably don't need the setTimeout since we're not triggering a focus anymore, should be fine to set the id immediately

File: src/app/shared/navbar/navbar.ts [hunk 29]
```diff
@@ -17,8 +19,21 @@ const SECTIONS_KEYS = Object.keys(SECTIONS);
...
+  private subscriptions = new Subscription();
   isNextVersion = location.hostname.startsWith('next.material.angular.io');
+  skipLinkHref: string;
+  skipLinkHidden = true;
+
+  constructor(router: Router) {
+    this.subscriptions.add(router.events
+      .pipe(filter((event: Event) => event instanceof NavigationEnd))
+      .subscribe(() => {
+        // set skip link
```
Comment (mmalerba) [inline src/app/shared/navbar/navbar.ts:32]: comment not useful, improve or remove

File: src/app/shared/navbar/navbar.ts [hunk 30]
```diff
@@ -17,8 +19,21 @@ const SECTIONS_KEYS = Object.keys(SECTIONS);
   templateUrl: './navbar.html',
   styleUrls: ['./navbar.scss']
 })
-export class NavBar {
+export class NavBar implements OnDestroy {
+  private subscriptions = new Subscription();
```
Comment (atscott) [inline src/app/shared/navbar/navbar.ts:23]: A common approach for unsubscribing is to have an onDestroy observable, i.e.
```
private readonly onDestroy = new Subject<void>();

...
router.events.pipe(takeUntil(this.onDestroy), ...).subscribe(...);
...

ngOnDestroy() {
  this.onDestroy.next();
  this.onDestroy.complete();
}
```

Up to you - this is more or less an FYI. It's more useful when you have multiple subscriptions that you need to unsubscribe from or as a project-wide convention.

Comment (atscott) [reply src/app/shared/navbar/navbar.ts:23]: I misread how this was being used. I like this `subscriptions.add` too :)

File: src/app/pages/component-viewer/component-viewer.ts (hunk 28 above)
Comment (annieyw) [reply src/app/pages/component-viewer/component-viewer.ts:92]: I was able to just remove this entirely because now `main-content` is set to the the tab headers.

**Comment by annieyw** [comment]
- remove focus outline 
- add `main-content` to individual guides page
- adjust skip link to land on tabs for individual component/cdk pages to allow easy nav to overview/api/examples

PTAL
